### PR TITLE
[297] Gave permission to rodsadmin and groupadmin to modify metadata …

### PR DESCRIPTION
**Pull request by connectbo**
@trel Alright I've got something for your review. This pr includes the following changes
- admins (rodsadmin and groupadmin) are able to modify the template, or set it as public/private if needed (not sure if we should allow this), and metalnx will also display the template owner in the tooltip
image

- changed read-only template form fields to be 'disabled', so it's not only greyed out but also can't be edited
image

So in short, admins are allowed to view/edit/apply all public templates and set templates as private but can't view or edit any private templates. Users are able to view/apply all public templates and edit their own templates.

Does that sound like something we can add to metalnx for now? Or we should wait till the next metadata template meeting?

### CODE REVIEW

File: src/metalnx-web/src/main/java/com/emc/metalnx/controller/TemplateController.java [hunk 1]
```diff
@@ -301,7 +301,9 @@ public String modifyTemplate(final Model model, @ModelAttribute final MetadataTe
 				throw new Exception("Cannot modify a non-existent template");
 			}
 			DataGridUser loggedUser = loggedUserUtils.getLoggedDataGridUser();
-			if (!template.getOwner().equalsIgnoreCase(loggedUser.getUsername())) {
+
+			// check if user is the template owner or admin
+			if (!(template.getOwner().equalsIgnoreCase(loggedUser.getUsername()) || loggedUser.isAdmin())) {
```
Comment (trel) [inline src/metalnx-web/src/main/java/com/emc/metalnx/controller/TemplateController.java:306]: so `loggedUser.isAdmin()` is either rodsadmin OR groupadmin?    those are not the same...

I think there are three cases we want to allow editing...

1) loggedUser is template creator/owner
2) loggedUser is rodsadmin
3) loggedUser is groupadmin of group that the template creator/owner is a part of

Comment (connectbo) [reply src/metalnx-web/src/main/java/com/emc/metalnx/controller/TemplateController.java:306]: oh I thought `loggedUser.isAdmin()` would return true if `groupadmin` as well.. thanks, will change that.

Comment (trel) [reply src/metalnx-web/src/main/java/com/emc/metalnx/controller/TemplateController.java:306]: Right, where does that `isAdmin()` code live?   I couldn't find it in a cursory grep...

Comment (trel) [reply src/metalnx-web/src/main/java/com/emc/metalnx/controller/TemplateController.java:306]: if `isAdmin()` does return true for groupadmin as well, I would argue that is a separate bug to file and to fix...

Comment (connectbo) [reply src/metalnx-web/src/main/java/com/emc/metalnx/controller/TemplateController.java:306]: nope, `isAdmin()` only returns true for rodsadmin, and it's inside the DataGridUser class, which is in the jargon layer. There's also an `isGroupAdmin()` available that checks the user type. I will add some code to check if this user is a groupadmin of that template owner's group.

Comment (trel) [reply src/metalnx-web/src/main/java/com/emc/metalnx/controller/TemplateController.java:306]: perfect.

File: src/metalnx-web/src/main/java/com/emc/metalnx/controller/TemplateController.java [hunk 2]
```diff
@@ -302,10 +312,27 @@ public String modifyTemplate(final Model model, @ModelAttribute final MetadataTe
...
+			boolean groupAdminEditPermission = false;
+			String[] templateOwnerGroupList = userService.getGroupIdsForUser(template.getOwner(), irodsServices.getCurrentUserZone());
+			Set<String> groupAdminGroupList = new HashSet<>(Arrays.asList(userService.getGroupIdsForUser(loggedUser.getUsername(), irodsServices.getCurrentUserZone())));
+			logger.info("groupList length is: " + templateOwnerGroupList.length);
+			for(int i = 0; i< templateOwnerGroupList.length && !groupAdminEditPermission; i++){
+				logger.info("current user is: " + templateOwnerGroupList[i]);
+				if(groupAdminGroupList.contains(templateOwnerGroupList[i])) groupAdminEditPermission = true;
+			}
+			if(!groupAdminEditPermission) throw new Exception("Cannot modify a template belonging to a user of other group");
+
```
Comment (trel) [inline src/metalnx-web/src/main/java/com/emc/metalnx/controller/TemplateController.java:330]: please indent 321-330

please break 327 and 329 into three lines each (using brackets to make the if blocks safer)

File: src/metalnx-web/src/main/java/com/emc/metalnx/controller/TemplateController.java [hunk 3]
```diff
@@ -75,6 +79,12 @@
 	@Autowired
 	UserService userService;
 
+	@Autowired
+	GroupService groupService;
```
Comment (trel) [inline src/metalnx-web/src/main/java/com/emc/metalnx/controller/TemplateController.java:83]: is this used?

**Comment by connectbo** [comment]
Just did some testing and cleaned up the code, think it's ready. Please review.

File: src/metalnx-web/src/main/java/com/emc/metalnx/controller/TemplateController.java [hunk 4]
```diff
@@ -301,8 +306,24 @@ public String modifyTemplate(final Model model, @ModelAttribute final MetadataTe
...
 			}
 			DataGridUser loggedUser = loggedUserUtils.getLoggedDataGridUser();
-			if (!template.getOwner().equalsIgnoreCase(loggedUser.getUsername())) {
-				throw new Exception("Cannot modify a template beloging to another user");
+
+			// throw error if user is not the template owner, rodsadmin or groupadmin
+ 			if (!(template.getOwner().equalsIgnoreCase(loggedUser.getUsername())) && loggedUser.isAdmin() == false && loggedUser.isGroupAdmin() == false) {
+				throw new Exception("Cannot modify a template belonging to another user");
+			}
+			// throw error if logged user is not the rodsadmin of the template owner's group
```
Comment (trel) [inline src/metalnx-web/src/main/java/com/emc/metalnx/controller/TemplateController.java:314]: in this comment...    rodsadmin -> groupadmin

**Comment by trel** [comment]
fix the one comment, and then # it up.   thanks!

File: src/metalnx-web/src/main/java/com/emc/metalnx/controller/TemplateController.java (hunk 4 above)
Comment (connectbo) [reply src/metalnx-web/src/main/java/com/emc/metalnx/controller/TemplateController.java]: yep, thanks for catching that
